sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala

```diff
@@ -1086,6 +1086,28 @@ object PruneFilters extends Rule[LogicalPlan] with PredicateHelper {
  * This heuristic is valid assuming the expression evaluation cost is minimal.
  */
 object PushDownPredicate extends Rule[LogicalPlan] with PredicateHelper {
+  /**
+   * Splits condition expressions into pushDown predicates and stayUp predicates based on
+   * specific rules. Parts of the predicate that can be pushed beneath must satisfy the following
+   * conditions:
+   * 1. Deterministic.
+   * 2. Placed before any non-deterministic predicates.
+   * 3. Other specific rules.
+   *
+   * @return (pushDown, stayUp)
+   */
+  private def splitPushdownPredicates(
+      condition: Expression)(specificRules: (Expression) => Boolean) = {
```

Comment:
hmmm, looks like duplicating these codes is more readable, @liancheng what do you think?